src/map/quest.cpp

```diff
@@ -709,32 +719,32 @@ void quest_update_objective(struct map_session_data *sd, struct mob_data* md)
 
 		// Process quest objectives
 		for (int j = 0; j < qi->objectives.size(); j++) {
-			uint8 objective_check = 0; // Must pass all 6 checks
-
-			if (qi->objectives[j]->mob_id == md->mob_id)
-				objective_check = 6;
-			else if (qi->objectives[j]->mob_id == 0) {
-				if (qi->objectives[j]->min_level == 0 || qi->objectives[j]->min_level <= md->level)
-					objective_check++;
-				if (qi->objectives[j]->max_level == 0 || qi->objectives[j]->max_level >= md->level)
-					objective_check++;
-				if (qi->objectives[j]->race == RC_ALL || qi->objectives[j]->race == md->status.race)
-					objective_check++;
-				if (qi->objectives[j]->size == SZ_ALL || qi->objectives[j]->size == md->status.size)
-					objective_check++;
-				if (qi->objectives[j]->element == ELE_ALL || qi->objectives[j]->element == md->status.def_ele)
-					objective_check++;
-				if (qi->objectives[j]->mapid < 0 || (qi->objectives[j]->mapid == sd->bl.m && md->spawn != nullptr))
-					objective_check++;
-				else if (qi->objectives[j]->mapid >= 0) {
-					struct map_data *mapdata = map_getmapdata(sd->bl.m);
-
-					if (mapdata->instance_id && mapdata->instance_src_map == qi->objectives[j]->mapid)
-						objective_check++;
+			if (sd->quest_log[i].count[j] < qi->objectives[j]->count)  {
+				if (qi->objectives[j]->mob_id != 0 && qi->objectives[j]->mob_id != md->mob_id)
+					continue;
+				if (qi->objectives[j]->mob_id == 0) {
+					if (qi->objectives[j]->min_level != 0 && qi->objectives[j]->min_level > md->level)
+						continue;
+					if (qi->objectives[j]->max_level != 0 && qi->objectives[j]->max_level < md->level)
+						continue;
+					if (qi->objectives[j]->race != RC_ALL && qi->objectives[j]->race != md->status.race)
+						continue;
+					if (qi->objectives[j]->size != SZ_ALL && qi->objectives[j]->size != md->status.size)
+						continue;
+					if (qi->objectives[j]->element != ELE_ALL && qi->objectives[j]->element != md->status.def_ele)
+						continue;
+					if (qi->objectives[j]->mapid >= 0) {
+						map_data *mapdata = map_getmapdata(sd->bl.m);
+
+						if (mapdata->instance_id < 1 && qi->objectives[j]->mapid != sd->bl.m)
+							continue;
+						if (mapdata->instance_id > 0 && mapdata->instance_src_map != qi->objectives[j]->mapid)
```

Comment:
```suggestion
						if (qi->objectives[j]->mapid != sd->bl.m && mapdata->instance_src_map != qi->objectives[j]->mapid)
```
Would it not work like this too?